src/bytes.rs

```diff
@@ -104,16 +105,51 @@ pub struct Bytes {
     data: AtomicPtr<()>,
     vtable: &'static Vtable,
 }
+/// A trait for underlying implementations for `Bytes` type.
+///
+/// All implementations must fulfill the following requirements:
+/// - They are cheaply cloneable and thereby shareable between an unlimited amount
+///   of components, for example by modifying a reference count.
+/// - Instances can be sliced to refer to a subset of the the original buffer.
```

Comment:
The trait and its methods all require a `# Safety` section in order to comply with the Repo's clippy config.

Reply:
Just to say this PR itself wasn't intended to be merged as-is, mostly because I was afraid it's too big to be merged at once. I planned to split it into multiple PRs once it got enough support. Also I want comments from the original author since the original logics are untouched and well, it's the `# Safety` section of the public trait document.

I checked your PR to this branch. It looks good for me except for the default clone impl, for the reason I've described in above comment. The default "do nothing on clone" strategy would be only valid for `&'static [u8]` buffers which is covered by builtin impl.

Reply:
Yeah.  That default impl was sort of a placeholder.  

I think we need to come up with a layered strategy, though.  
We should offer a default refcounted implementation for those that don't care about:
1. Don't care about copy-on-write optimizations
2. Special rules for cloning,